buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy

```diff
@@ -103,16 +103,76 @@ class BuildPlugin implements Plugin<Project> {
 
     /** Makes dependencies non-transitive by default */
     static void configureConfigurations(Project project) {
+        // fail on any conflicting dependency versions
+        project.configurations.all({ Configuration configuration ->
+            if (configuration.name.startsWith('_transitive_')) {
+                // don't force transitive configurations to not conflict with themselves, since
+                // we just have them to find *what* transitive deps exist
+                return
+            }
+            configuration.resolutionStrategy.failOnVersionConflict()
+        })
 
         // force all dependencies added directly to compile/testCompile to be non-transitive, except for ES itself
-        project.configurations.compile.dependencies.all { dep ->
+        Closure disableTransitiveDeps = { ModuleDependency dep ->
             if (!(dep instanceof ProjectDependency) && dep.getGroup() != 'org.elasticsearch') {
                 dep.transitive = false
+
+                // also create a configuration just for this dependency version, so that later
+                // we can determine which dependencies it has (for pom generation excludes)
```

Comment:
The leap here is that the configuration groups a direct dependency with all of its own transitive dependencies for easy checking later. That isn't clear unless you've worked your brain in that special maven + gradle way.